[Security Solution] Webhook - Case Management Connector Documentation (#137726)
elastic/kibana: docs/management/connectors/action-types/cases-webhook.asciidoc

[role="xpack"]
[[cases-webhook-action-type]]
=== Webhook - Case Management connector and action
++++
<titleabbrev>Webhook - Case Management</titleabbrev>
++++

The Webhook - Case Management connector uses https://github.com/axios/axios[axios] to send POST, PUT, and GET requests to a Case Management RESTful API web service. The resulting connector can be used with Kibana Cases to create and update a case in the third party web service.

[float]
[[webhook-connector-configuration]]
==== Connector configuration

Webhook - Case Management connectors have the following configuration properties.

Name::      The name of the connector. The name is used to identify a  connector in the management UI connector listing, or in the connector list when configuring an action.
Require authentication:: If true, a username and password for login type authentication must be provided.
Username::      Username for HTTP basic authentication.
Password::  Password for HTTP basic authentication.

> Headers::   A set of key-value pairs sent as headers with the request

Reviewer: I think it is valuable to mention, that this headers will be added to each of the URLs for create, update and comment.

Create Case Method::  REST API HTTP request method to create case in third party system, either `post`(default), `put`, or `patch`.
Create Case URL::       Rest API URL to create case in third party system. If you are using the <<action-settings, `xpack.actions.allowedHosts`>> setting, make sure the hostname is added to the allowed hosts.
Create Case Object::      A JSON payload sent to the create case URL to create case. Use the variable selector to add Kibana Cases data to the payload. Required variables are `case.title` and `case.description`. For example:
+
[source,text]
--
{
	"fields": {
	  "summary": {{{case.title}}},
	  "description": {{{case.description}}},
	  "labels": {{{case.tags}}}
	}
}
--
Note: Due to Mustache template variables (the text enclosed in triple braces, for example, `{{{case.title}}}`), the JSON is not validated in this step. The JSON is validated once the mustache variables have been placed when REST method is executed. Manually ensure that the JSON is valid besides the mustache variables so the later validation will pass.

> Create Case Response - Case ID Key::   JSON key in create case response that contains the external case id

Reviewer: I suggest to have a working example with the real configuration to help the user to understand better the capabilities of the new connector.
It could be some existing integration like ServiceNow or Jira. The service endpoint won't be the real thing, but all the parameters and requests bodies will be exactly what the real integration have. Only thin the user will need to do is to follow the suggested steps but replace the service endpoints with their own.

Author: i kind of followed the other documentation. should i do this in a new section, or break from the model of other docs and do it inline?

Reviewer: You are right, the current doc follows the structure of the other connectors. But having the section example looks like a great idea!
Also, @elastic/security-docs maybe you can suggest something here?

Author: I've attempted an example in my [last commit](https://github.com/elastic/kibana/pull/137726/commits/214745a51158e11eec955370fed854dd00ead01b)

`viewIncidentUrl`:: A URL string with either the External Service Id or External Service Title mustache variable that corresponds to *View Case URL*.
`updateIncidentUrl`:: A URL string that corresponds to *Update Case URL*.
`updateIncidentMethod`:: A string that corresponds to *Update Case Method*.
`updateIncidentJson`:: A stringified JSON with mustache variables that corresponds to *Update Case JSON*.
`createCommentUrl`:: A URL string that corresponds to *Create Comment URL*.
`createCommentMethod`:: A string that corresponds to *Create Comment Method*.
`createCommentJson`:: A stringified JSON with mustache variables that corresponds to *Create Comment JSON*.

Secrets defines sensitive information for the connector type.

`user`:: A string that corresponds to *User*. Required if `hasAuth` is set to `true`.
`password`:: A string that corresponds to *Password*. Required if `hasAuth` is set to `true`.

[float]
[[define-webhook-ui]]
==== Define connector in Stack Management

Define Webhook - Case Management connector properties.

[role="screenshot"]

> image::management/connectors/images/cases-webhook-connector.gif[Webhook - Case Management connector]

Reviewer: It will be more clear if this gifs will correspond to the example steps.

Author: like you mean the same values? i was trying to create a new gif but my software keeps freezing...

Reviewer: I can help you with recording the example gifs and URLs/bodies , I have it is working somewhere on my remote cluster.

++++
<titleabbrev>Webhook - Case Management</titleabbrev>
++++

The Webhook - Case Management connector uses https://github.com/axios/axios[axios] to send POST, PUT, and GET requests to a Case Management RESTful API web service. The resulting connector can be used with Kibana Cases to create and update a case in the third party web service.

[float]
[[cases-webhook-connector-configuration]]
=== Connector configuration

Webhook - Case Management connectors have the following configuration properties.

Name::      The name of the connector. The name is used to identify a  connector in the management UI connector listing, or in the connector list when configuring an action.
Require authentication:: If true, a username and password for login type authentication must be provided.
Username::      Username for HTTP basic authentication.
Password::  Password for HTTP basic authentication.
Headers::   A set of key-value pairs sent as headers with the request URLs for the create case, update case, get case, and create comment methods.
Create Case Method::  REST API HTTP request method to create case in third party system, either `post`(default), `put`, or `patch`.

> Create Case URL::       Rest API URL to create case in third party system. If you are using the <<action-settings, `xpack.actions.allowedHosts`>> setting, make sure the hostname is added to the allowed hosts.

Author: i dont think this should be `host.name`, since we're not talking about ECS but instead a website domain name

`createIncidentJson`:: A stringified JSON with mustache variables that corresponds to *Create Case JSON*.
`createIncidentResponseKey`:: A string from the response body of the create case method that corresponds to the *External Service Id*.
`getIncidentUrl`:: A URL string with External Service Id mustache variable that corresponds to *Get Case URL*.
`getIncidentResponseExternalTitleKey`:: A string from the response body of the get case method that corresponds to the *External Service Title*.
`viewIncidentUrl`:: A URL string with either the External Service Id or External Service Title mustache variable that corresponds to *View Case URL*.
`updateIncidentUrl`:: A URL string that corresponds to *Update Case URL*.
`updateIncidentMethod`:: A string that corresponds to *Update Case Method*.
`updateIncidentJson`:: A stringified JSON with mustache variables that corresponds to *Update Case JSON*.
`createCommentUrl`:: A URL string that corresponds to *Create Comment URL*.
`createCommentMethod`:: A string that corresponds to *Create Comment Method*.
`createCommentJson`:: A stringified JSON with mustache variables that corresponds to *Create Comment JSON*.

> Secrets defines sensitive information for the connector type.

Author: this is in every connector doc, labeling the config and secrets fields
<img width="1169" alt="Screen Shot 2022-08-08 at 8 06 42 AM" src="https://user-images.githubusercontent.com/6935300/183437191-f4485ffa-b228-4ca3-95fc-cd953293d1b8.png">

`user`:: A string that corresponds to *User*. Required if `hasAuth` is set to `true`.
`password`:: A string that corresponds to *Password*. Required if `hasAuth` is set to `true`.

[float]
[[define-cases-webhook-ui]]
=== Define connector in Stack Management

> Define Webhook - Case Management connector properties.

Author: Well they can also do it directly in the cases workflow, Cases > Configure

[role="screenshot"]
image::management/connectors/images/cases-webhook-connector.gif[Webhook - Case Management connector]

Test Webhook - Case Management action parameters.

[role="screenshot"]
image::management/connectors/images/cases-webhook-test.gif[Webhook - Case Management params test]

[float]
[[cases-webhook-action-configuration]]
=== Action configuration

Webhook - Case Management actions have the following configuration properties.

Title:: A title for the issue, used for searching the contents of the knowledge base.
Description:: The details about the incident.
Labels:: The labels for the incident.
Additional comments:: Additional information for the client, such as how to troubleshoot the issue.

[float]
[[cases-webhook-connector-full-example]]

> == Full Example with Third Party System

Reviewer 2: ```suggestion
== Example: Setting up a the Webhook - Case Management Connector with {jira} 
```

Author: does this make a link to jira?

The following is an example of using the Webhook - Case Management Connector with a real third party case management system, Jira. We will reference their https://developer.atlassian.com/cloud/jira/platform/rest/v2/api-group-issues/[API documentation] for creating an issue and use a demo instance of Jira with real working endpoints.

_Note: We are using Jira as an example third party. However, if you wish to connect with Jira, using the <<jira-action-type, preconfigured Jira connector>> will simplify your setup._

[float]

> ====  Step 1 - Set up connector

Reviewer 2: ```suggestion
====  Set up connector
```

Reviewer 2: Matching this with the text that's in the flyout so it's exactly the same.

Author: <img width="652" alt="Screen Shot 2022-08-08 at 8 10 40 AM" src="https://user-images.githubusercontent.com/6935300/183437997-1b45f24d-3a42-46dc-9350-fcc1f03fcf62.png">
i think we should leave the step numbers in, they are part of the flyout text

> In the Webhook - Case Management connector create flyout, we begin by entering a *Name* for the connector - `Jira Test Connector`. We are going to use basic authentication in this example, so we keep the *Require authentication* checkbox checked and enter the *Username* and *Password* for the test instance, `test-user@elastic.co` and `notarealpassword`. We will not be setting any *Headers* for the requests.

Reviewer 2: ```suggestion
In the Webhook - Case Management connector create flyout, begin by entering a connector *Name*, for example, `Jira Test Connector`. Basic authentication will be used in this example, so keep the *Require authentication* option selected and enter the *Username* and *Password* for the test instance, for example, `test-user@elastic.co` and `notarealpassword`. Headers will not be set for requests. 
```

[role="screenshot"]
image::management/connectors/images/cases-webhook-step1.png[Webhook - Case Management connector Step 1, Jira example]

[float]

> ====  Step 2 - Create case

Reviewer 2: ```suggestion
====  Create case
```

> Taking a look at the Jira https://developer.atlassian.com/cloud/jira/platform/rest/v2/api-group-issues/#api-rest-api-2-issue-post[Create issue method documentation], we can find what values we will need to fill step 2.

Reviewer 2: ```suggestion
To find the required values for this step, refer to the {jira} https://developer.atlassian.com/cloud/jira/platform/rest/v2/api-group-issues/#api-rest-api-2-issue-post[Create issue method documentation].
```

Reviewer 2: It might also be helpful to point towards where the information is in the doc. That way user aren't immediately able to ID what they should be looking for. :)

Also, I like what you're doing with walking users though each step in the example. One way we could continue to do that here is to describe the purpose of this step. Is it to create a case record in JIRA once the connector is created? Or is it to set up the connector so that a case is created in JIRA once users push changes to JIRA? Or is it something entirely different?

> Jira create issue request method: `POST`

Reviewer 2: If possible, I'd put this and what's on lines 199 and 201 in paragraph form. That'd keep it consistent with the tone and format you're using throughout the example.

Author: I repeat this format for each method/url. for example lines 240, 242, 331, 333, 335, 355, 357, 359

> Jira create issue request method: `POST`

Reviewer 2: ```suggestion
{jira} create issue request method: `POST`
```

> Jira create issue request url: `/rest/api/2/issue`

Reviewer 2: ```suggestion
{jira} create issue request url: `/rest/api/2/issue`
```

> Jira create issue request body:

Reviewer 2: ```suggestion
{jira} create issue request body:
```

[source,json]
--
{
    "fields": {
        "summary": "Main order flow broken",
        "description": "Order entry fails when selecting supplier.",
        "labels": ["bugfix",  "blitz_test"],
        "project":{"key":"PROJ-123"},
        "issuetype":{"id":"10000"}
    }
}
--

Jira create issue response body:
[source,json]
--
{
  "id": "10000",
  "key": "ED-24",
  "self": "https://your-domain.atlassian.net/rest/api/2/issue/10000",
  "transition": {
    "status": 200,
    "errorCollection": {
      "errorMessages": [],
      "errors": {}
    }
  }
}
--

> In the screen capture below, on Step 2 we enter `POST` as the *Create Case Method* and `https://testing-jira.atlassian.net/rest/api/2/issue` as the **Create Case Url**. In our example Jira instance, the project key is "ROC" and the issuetype id is "10024". We have entered the Jira request JSON as the *Create Case Object*, updating the project key to "ROC" and the issuetype id to "10024". We then use the Case variable selector to enter where we will map the Kibana case title, Kibana case description, and Kibana case tags. The Jira response body contains an id with the JSON key of "id", so we enter `id` as the *Create Case Response - Case ID Key*.

Reviewer 2: ```suggestion
In the following screen capture, we enter `POST` as the *Create Case Method* and `https://testing-jira.atlassian.net/rest/api/2/issue` as the **Create Case Url**. In our example {jira} instance, the project key is "ROC" and the issuetype id is "10024". We have entered the {jira} request JSON as the *Create Case Object*, updating the project key to "ROC" and the issuetype id to "10024". We then use the Case variable selector to enter where we will map the Kibana case title, Kibana case description, and Kibana case tags. The {jira} response body contains an ID with the JSON key of "id", so we enter `id` as the *Create Case Response - Case ID Key*.
```

[role="screenshot"]
image::management/connectors/images/cases-webhook-step2.gif[Webhook - Case Management connector Step 2, Jira example]

[float]

> ====  Step 3 - Get case information

Reviewer 2: ```suggestion
====  Get case information
```

> Next we'll need to look at Jira's https://developer.atlassian.com/cloud/jira/platform/rest/v2/api-group-issues/#api-rest-api-2-issue-issueidorkey-[Get issue method documentation] to find the values for step 3. In the GET response JSON below, we thinned out some null/unrelated data so that we can focus on the fields we need.

Reviewer 2: ```suggestion
Next we'll need to look at {jira}'s https://developer.atlassian.com/cloud/jira/platform/rest/v2/api-group-issues/#api-rest-api-2-issue-issueidorkey-[Get issue method documentation] to find the values for this step. In the GET response JSON below, we thinned out some null and unrelated data so that we can focus on the fields we need.
```

> Jira get issue request url: `/rest/api/2/issue/{issueIdOrKey}`

Reviewer 2: ```suggestion
{jira} get issue request url: `/rest/api/2/issue/{issueIdOrKey}`
```

> Jira get issue response body:

Reviewer 2: ```suggestion
{jira} get issue response body:
```

                "id": 2,
                "key": "new",
                "colorName": "blue-gray",
                "name": "To Do"
            }
        },
        "description": "Kibana Description",
        "summary": "Kibana Title",
        "creator": {
            "self": "https://testing-jira.atlassian.net/rest/api/2/user?accountId=12345",
            "accountId": "12345",
            "emailAddress": "test-user@elastic.co",
            "displayName": "MLR-QA",
            "active": true,
            "timeZone": "Europe/Athens",
            "accountType": "atlassian"
        },
        "reporter": {
            "self": "https://testing-jira.atlassian.net/rest/api/2/user?accountId=12345",
            "accountId": "12345",
            "emailAddress": "test-user@elastic.co",
            "displayName": "MLR-QA",
            "active": true,
            "timeZone": "Europe/Athens",
            "accountType": "atlassian"
        },
        "comment": {
            "comments": [],
            "self": "https://testing-jira.atlassian.net/rest/api/2/issue/71964/comment",
            "maxResults": 0,
            "total": 0,
            "startAt": 0
        }
    }
}
--

> To make the Get Case URL, we see we need `/rest/api/2/issue/{issueIdOrKey}`. We will fill in the value with the issue id, which we stored in the last step as *Create Case Response - Case ID Key*. Using the variable selector on the Get Case URL input, we can see the issue id is stored as a mustache value `{{{external.system.id}}}`. So our value for *Get Case URL* will be `https://testing-jira.atlassian.net/rest/api/2/issue/{{{external.system.id}}}`.

Reviewer 2: ```suggestion
To make the Get Case URL, we need `/rest/api/2/issue/{issueIdOrKey}`. We will fill in the value with the issue ID, which we stored in the last step as *Create Case Response - Case ID Key*. Using the variable selector on the Get Case URL input, we can see the issue ID is stored as a mustache value `{{{external.system.id}}}`. So our value for *Get Case URL* will be `https://testing-jira.atlassian.net/rest/api/2/issue/{{{external.system.id}}}`.
```

In the response JSON we can see the title of the case is "ROC-538". The key for this value is `key` so we enter `key` as the *Get Case Response External Title Key* value.

We also need the *External Case View URL*. https://support.atlassian.com/jira-software-cloud/docs/link-an-issue/[Jira's documentation] instructs you to get the link from the issue itself. The format for this link looks like `https://<user’s subdomain>.atlassian.net/browse/<issueKey>`. We mapped `key` to the *Get Case Response External Title Key* and using the variable selector on the *External Case View URL* input, we can see the issue key is stored as a mustache value `{{{external.system.title}}}`. Using this, the value for *External Case View URL* is `https://testing-jira.atlassian.net/browse/{{{external.system.title}}}`.
[role="screenshot"]
image::management/connectors/images/cases-webhook-step3.gif[Webhook - Case Management connector Step 3, Jira example]

[float]

> ====  Step 4 - Comments and updates

Reviewer 2: ```suggestion
====  Comments and updates
```

> In Step 4 we need to set the REST API data for updates and comments. Let's look at Jira's https://developer.atlassian.com/cloud/jira/platform/rest/v2/api-group-issues/#api-rest-api-2-issue-issueidorkey-put[Edit issue documentation].

Reviewer 2: ```suggestion
During this step, we need to set the REST API data for updates and comments. Let's look at {jira}'s https://developer.atlassian.com/cloud/jira/platform/rest/v2/api-group-issues/#api-rest-api-2-issue-issueidorkey-put[Edit issue documentation].
```

> Jira update issue request method: `PUT`

Reviewer 2: ```suggestion
{jira} update issue request method: `PUT`
```

> Jira update issue request url: `/rest/api/2/issue/{issueIdOrKey}`

Reviewer 2: ```suggestion
{jira} update issue request url: `/rest/api/2/issue/{issueIdOrKey}`
```

> Jira update issue request body:

Reviewer 2: ```suggestion
{jira} update issue request body:
```

[source,json]
--
{
    "fields": {
        "summary": "Main order flow broken",
        "description": "Order entry fails when selecting supplier.",
        "labels": ["bugfix",  "blitz_test"],
        "project":{"key":"PROJ-123"},
        "issuetype":{"id":"10000"}
    }
}
--

> In the screen capture below, on Step 4 we enter `PUT` as the *Update Case Method* and `https://testing-jira.atlassian.net/rest/api/2/issue/{{{external.system.id}}}` as the **Update Case Url** using the variable selector to insert the `{{{external.system.id}}}`. Just like the create case JSON, have entered the Jira request JSON as the *Update Case Object*, updating the project key to "ROC" and the issuetype id to "10024". We then use the Case variable selector to enter where we will map the Kibana case title, Kibana case description, and Kibana case tags.

Reviewer 2: ```suggestion
In the following screen capture, we enter `PUT` as the *Update Case Method* and `https://testing-jira.atlassian.net/rest/api/2/issue/{{{external.system.id}}}` as the **Update Case Url** using the variable selector to insert the `{{{external.system.id}}}`. Just like the create case JSON, have entered the {jira} request JSON as the *Update Case Object*, updating the project key to "ROC" and the issuetype id to "10024". We then use the case variable selector to enter where we will map the case title, case description, and case tags.
```

[role="screenshot"]
image::management/connectors/images/cases-webhook-step4a.gif[Webhook - Case Management connector Step 4 Update, Jira example]

> Lastly we will look at Jira's https://developer.atlassian.com/cloud/jira/platform/rest/v2/api-group-issue-comments/#api-rest-api-2-issue-issueidorkey-comment-post[Add comment documentation] to fill out the optional comment REST fields.

Reviewer 2: ```suggestion
Lastly we will look at {jira}'s https://developer.atlassian.com/cloud/jira/platform/rest/v2/api-group-issue-comments/#api-rest-api-2-issue-issueidorkey-comment-post[Add comment documentation] to fill out the optional comment REST fields.
```

> Jira create comment request method: `POST`

Reviewer 2: ```suggestion
{jira} create comment request method: `POST`
```

> Jira create comment request url: `/rest/api/2/issue/{issueIdOrKey}/comment`

Reviewer 2: ```suggestion
{jira} create comment request url: `/rest/api/2/issue/{issueIdOrKey}/comment`
```

> Jira create comment request body:

Reviewer 2: ```suggestion
{jira} create comment request body:
```

[source,json]
--
{
    "body": "Lorem ipsum dolor sit amet."
}
--

> In the screen capture below, on Step 4 we enter `POST` as the *Create Comment Method* and `https://testing-jira.atlassian.net/rest/api/2/issue/{{{external.system.id}}}/comment` as the **Create Comment Url** using the variable selector to insert the `{{{external.system.id}}}`. We enter the Jira request JSON as the *Create Comment Object*, using the Case variable selector to enter where we will map the Kibana case comment.

Reviewer 2: ```suggestion
In the following screen capture, we enter `POST` as the *Create Comment Method* and `https://testing-jira.atlassian.net/rest/api/2/issue/{{{external.system.id}}}/comment` as the **Create Comment Url** using the variable selector to insert the `{{{external.system.id}}}`. We enter the {jira} request JSON as the *Create Comment Object*, using the case variable selector to enter where we will map the case comment.
```

[role="screenshot"]
image::management/connectors/images/cases-webhook-step4b.gif[Webhook - Case Management connector Step 4 Comments, Jira example]

[float]
[[cases-webhook-example-implementation]]
=== Implement connector in Kibana Cases

> Let's take a look at how our new Webhook - Case Management connector works within the Kibana Case Workflow.

Reviewer 2: ```suggestion
Let's take a look at how our new Webhook - Case Management connector works within the case workflow.
```

`viewIncidentUrl`::: A URL string with either the *External Service Id* or *External Service Title* Mustache variable that corresponds to *View Case URL*.
`updateIncidentUrl`::: A URL string that corresponds to *Update Case URL*.
`updateIncidentMethod`::: A string that corresponds to *Update Case Method*.
`updateIncidentJson`::: A stringified JSON with Mustache variables that corresponds to *Update Case JSON*.
`createCommentUrl`::: A URL string that corresponds to *Create Comment URL*.
`createCommentMethod`::: A string that corresponds to *Create Comment Method*.
`createCommentJson`::: A stringified JSON with Mustache variables that corresponds to *Create Comment JSON*.

`secrets`:: Defines sensitive information for the connector type.
`user`::: A string that corresponds to *User*. Required if `hasAuth` is set to `true`.
`password`::: A string that corresponds to *Password*. Required if `hasAuth` is set to `true`.

[float]
[[define-cases-webhook-ui]]
=== Define connector in {stack-manage-app}

Define Webhook - Case Management connector properties:

[role="screenshot"]

> image::management/connectors/images/cases-webhook-connector.gif[Webhook - Case Management connector]

Reviewer 3: > image::management/connectors/images/cases-webhook-connector.gif[Webhook - Case Management connector]

IMO this would be simpler to maintain as a static image, like in the other connector pages. Ideally the UI is intuitive enough that it doesn't require this animation.

Test Webhook - Case Management action parameters:

[role="screenshot"]

> image::management/connectors/images/cases-webhook-test.gif[Webhook - Case Management params test]

Reviewer 3: > image::management/connectors/images/cases-webhook-test.gif[Webhook - Case Management params test]

Ditto re preference for static image that we can then automatically update and maintain.

[float]
[[cases-webhook-action-configuration]]
=== Action configuration

Webhook - Case Management actions have the following configuration properties:

Title:: A title for the issue, which is used for searching the contents of the knowledge base.
Description:: The details about the incident.
Labels:: The labels for the incident.
Additional comments:: Additional information for the client, such as how to troubleshoot the issue.

[float]
[[cases-webhook-connector-full-example]]

> == Full example with third-party system

Reviewer 3: > == Full example with third-party system

IMO everything hereafter is more suited to a blog rather than the docs. These connector docs are intended as reference information (and will gradually be reduced even further as we move the information into the appropriate API reference pages).  So an example like this (especially since it says the preconfigured Jira connector is the preferred choice) seems more like a learning exercise in a blog.

Author: my tech lead requested we put an example here. if you want it in a blog, can we publish it at the same time this is published and link it?

Reviewer 3: Yes, we have links to important blogs in the Machine Learning docs, so it is fine to do so here too once that information is published.
